src/sdam/srv_polling/mod.rs

```diff
@@ -0,0 +1,139 @@
+#[cfg(test)]
+mod test;
+
+use std::time::Duration;
+
+use super::{
+    monitor::DEFAULT_HEARTBEAT_FREQUENCY,
+    state::{Topology, TopologyState, WeakTopology},
+};
+use crate::{
+    error::{Error, Result},
+    options::{ClientOptions, StreamAddress},
+    srv::SrvResolver,
+    RUNTIME,
+};
+
+const DEFAULT_RESCAN_SRV_INTERVAL: Duration = Duration::from_secs(60);
+
+pub(crate) struct SrvPollingMonitor {
+    initial_hostname: String,
+    resolver: Option<SrvResolver>,
+    topology: WeakTopology,
+    rescan_interval: Option<Duration>,
+    client_options: ClientOptions,
+}
+
+impl SrvPollingMonitor {
+    pub(crate) fn new(topology: WeakTopology) -> Option<Self> {
+        let client_options = topology.client_options().clone();
+
+        let initial_hostname = match client_options.original_srv_hostname() {
+            Some(hostname) => hostname.clone(),
+            None => return None,
+        };
+
+        Some(Self {
+            initial_hostname,
+            resolver: None,
+            topology,
+            rescan_interval: None,
+            client_options,
+        })
+    }
+
+    /// Starts a monitoring task that periodically performs SRV record lookups to determine if the
+    /// set of mongos in the cluster have changed. A weak reference is used to ensure that the
+    /// monitoring task doesn't keep the topology alive after the client has been dropped.
+    pub(super) fn start(topology: WeakTopology) {
+        RUNTIME.execute(async move {
+            if let Some(mut monitor) = Self::new(topology) {
+                monitor.execute().await;
+            }
+        });
+    }
+
+    async fn execute(&mut self) {
+        while let Some(topology) = self.topology.upgrade() {
+            let state = topology.clone_state().await;
+
+            if state.is_sharded() || state.is_unknown() {
+                let hosts = self.lookup_hosts().await;
+                self.update_hosts(hosts, topology, state).await;
+            }
+
+            RUNTIME
+                .delay_for(self.rescan_interval.unwrap_or(DEFAULT_RESCAN_SRV_INTERVAL))
+                .await;
+        }
+    }
+
+    pub(crate) async fn update_hosts(
+        &mut self,
+        hosts: Result<Vec<StreamAddress>>,
+        topology: Topology,
+        mut topology_state: TopologyState,
+    ) {
+        let hosts = match hosts {
+            Ok(hosts) if hosts.is_empty() => {
+                self.no_valid_hosts(None);
+
+                return;
+            }
+            Ok(hosts) => hosts,
+            Err(err) => {
+                self.no_valid_hosts(Some(err));
+
+                return;
+            }
+        };
+
+        // TODO: Use TTL values to determine delay duration once `trust-dns-resolver`
```

Comment:
As mentioned in the slack channel, trust-dns-resolver added the functionality to be able to inspect the TTL values of the SRV records right after their last release, but they haven't released since then, so we can't implement the optimization to hold off on future checks until the TTL has expired.

Reply:
sounds good, can we file a ticket for that and update this todo to reference it?